addons/binding/org.openhab.binding.connectsdk/src/main/java/org/openhab/binding/connectsdk/handler/ConnectSDKHandler.java

```diff
@@ -0,0 +1,235 @@
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.State;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.connectsdk.device.ConnectableDevice;
+import com.connectsdk.device.ConnectableDeviceListener;
+import com.connectsdk.discovery.DiscoveryManager;
+import com.connectsdk.discovery.DiscoveryManagerListener;
+import com.connectsdk.service.DeviceService;
+import com.connectsdk.service.DeviceService.PairingType;
+import com.connectsdk.service.command.ServiceCommandError;
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * The {@link ConnectSDKHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Sebastian Prehn - Initial contribution
+ */
+public class ConnectSDKHandler extends BaseThingHandler implements ConnectableDeviceListener, DiscoveryManagerListener {
+
+    private Logger logger = LoggerFactory.getLogger(ConnectSDKHandler.class);
+    private DiscoveryManager discoveryManager;
+
+    // ChannelID to CommandHandler Map
+    private final Map<String, ChannelHandler> channelHandlers = ImmutableMap.<String, ChannelHandler>builder()
+            .put(CHANNEL_VOLUME, new VolumeControlVolume()).put(CHANNEL_VOLUME_UP, new VolumeControlUp())
+            .put(CHANNEL_VOLUME_DOWN, new VolumeControlDown()).put(CHANNEL_POWER, new PowerControlPower())
+            .put(CHANNEL_MUTE, new VolumeControlMute()).put(CHANNEL_CHANNEL, new TVControlChannel())
+            .put(CHANNEL_CHANNEL_UP, new TVControlUp()).put(CHANNEL_CHANNEL_DOWN, new TVControlDown())
+            .put(CHANNEL_CHANNEL_NAME, new TVControlChannelName()).put(CHANNEL_APP_LAUCHER, new LauncherApplication())
+            .put(CHANNEL_MEDIA_FORWARD, new MediaControlForward()).put(CHANNEL_MEDIA_PAUSE, new MediaControlPause())
+            .put(CHANNEL_MEDIA_PLAY, new MediaControlPlay()).put(CHANNEL_MEDIA_REWIND, new MediaControlRewind())
+            .put(CHANNEL_MEDIA_STOP, new MediaControlStop()).put(CHANNEL_MEDIA_STATE, new MediaControlPlayState())
+            .put(CHANNEL_TOAST, new ToastControlToast()).build();
+
+    public ConnectSDKHandler(Thing thing, DiscoveryManager discoveryManager) {
+        super(thing);
+        this.discoveryManager = discoveryManager;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.debug("internalReceiveCommand({},{}) is called", channelUID, command);
+        ChannelHandler handler = channelHandlers.get(channelUID.getId());
+        final ConnectableDevice d = getDevice();
+        if (d == null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    String.format("%s not found under connect sdk devices", getThing().getUID()));
+            return;
+        }
+        handler.onReceiveCommand(d, command);
```

Comment:
Can this even happen? I assume as long as channel Id in thing-types.xml and channelHandlers keys match, no. So it would be a programming bug. If this would occur it would cause a NullPointerException.

Reply:
It should not happen and would mean an error, so we could log an error. This would be more graceful than to throw an exception.

Reply:
Done